{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "from torch_geometric.loader import DataLoader\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.io import mmread\n",
    "import matplotlib.pyplot as plt\n",
    "import anndata as ad\n",
    "import h5py\n",
    "import scanpy as sc\n",
    "import re\n",
    "import os\n",
    "import os.path as osp\n",
    "import sys\n",
    "from matplotlib_venn import venn2, venn3\n",
    "from random import sample\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import networkx as nx\n",
    "from torch_geometric.data import Data\n",
    "from torch_geometric.utils import add_remaining_self_loops, to_undirected\n",
    "from tqdm import tqdm\n",
    "import torch.optim as optim\n",
    "from pdgrapher import Dataset, PDGrapher, Trainer\n",
    "\n",
    "# borrowed from stackoverflow (https://stackoverflow.com/questions/61133916/is-there-in-python-a-single-function-that-shows-the-full-structure-of-a-hdf5-fi)\n",
    "def h5_tree(val, pre=''):\n",
    "    items = len(val)\n",
    "    for key, val in val.items():\n",
    "        items -= 1\n",
    "        if items == 0:\n",
    "            # the last item\n",
    "            if type(val) == h5py._hl.group.Group:\n",
    "                print(pre + '└── ' + key)\n",
    "                h5_tree(val, pre+'    ')\n",
    "            else:\n",
    "                try:\n",
    "                    print(pre + '└── ' + key + ' (%d)' % len(val))\n",
    "                except TypeError:\n",
    "                    print(pre + '└── ' + key + ' (scalar)')\n",
    "        else:\n",
    "            if type(val) == h5py._hl.group.Group:\n",
    "                print(pre + '├── ' + key)\n",
    "                h5_tree(val, pre+'│   ')\n",
    "            else:\n",
    "                try:\n",
    "                    print(pre + '├── ' + key + ' (%d)' % len(val))\n",
    "                except TypeError:\n",
    "                    print(pre + '├── ' + key + ' (scalar)')\n",
    "\n",
    "def output_divider():\n",
    "    print('----------------------------------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4/4\n",
    "\n",
    "Starting today our objective shifts to using LINCS level 3 overexpression data. LINCS is a bulk, microarray seqenced dataset. Julia's data is single cell and could be a problem due to mainly inherent distribution difference from the training dataset. \n",
    "\n",
    "Today we aim to download and QC the oe dataset using the same script for xpr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wget https://s3.amazonaws.com/macchiato.clue.io/builds/LINCS2020/level3/level3_beta_trt_oe_n131668x12328.gctx -O /home/b-evelyntong/hl/lincs_lvl3_oe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "### pushing through every line of code that processes the data from its original format into npz format\n",
    "DATA_ROOT=\"/home/b-evelyntong/hl/lincs_lvl3_oe\"\n",
    "\n",
    "### STEP 1 LOAD\n",
    "# this chunk loads in instinfo_beta and geneinfo_beta, specific to LINCS, the former for cells and latter for information on genes\n",
    "# instinfo contains information on which perturbation each cell went through\n",
    "inst_info = pd.read_csv(os.path.join(DATA_ROOT, 'instinfo_beta.txt'), sep=\"\\t\", low_memory=False)\n",
    "gene_info = pd.read_csv(os.path.join(DATA_ROOT, 'geneinfo_beta.txt'), sep=\"\\t\", low_memory=False)\n",
    "\n",
    "# these steps make sure that inst_info_xxx contains the subset of cells that are either is in the \n",
    "# treatment group and perturbed with crispr, or is in the control group and received only a vector\n",
    "inst_info_ctl = inst_info[np.logical_and(inst_info['pert_type'] == 'ctl_vector', inst_info['failure_mode'].isna()) ].reset_index(inplace=False, drop=True)\n",
    "inst_info_oe = inst_info[np.logical_and(inst_info['pert_type'] == 'trt_oe', inst_info['failure_mode'].isna()) ].reset_index(inplace=False, drop=True)\n",
    "\n",
    "# reads in h5py file and creates a panda dataframe out of it\n",
    "f = h5py.File(os.path.join(DATA_ROOT, 'level3_beta_trt_oe_n131668x12328.gctx'), 'r')\n",
    "matrix_oe = f['0']['DATA']['0']['matrix'][:].transpose()\n",
    "gene_ids_oe = f['0']['META']['ROW']['id'][:]\n",
    "sample_ids_oe = f['0']['META']['COL']['id'][:]\n",
    "matrix_oe = pd.DataFrame(matrix_oe, columns = sample_ids_oe.astype(str), index = gene_ids_oe.astype(int))\n",
    "del f\n",
    "\n",
    "### Control data\n",
    "f = h5py.File(os.path.join(DATA_ROOT, 'level3_beta_ctl_n188708x12328.gctx'), 'r')\n",
    "matrix_ctl = f['0']['DATA']['0']['matrix'][:].transpose()\n",
    "gene_ids_ctl = f['0']['META']['ROW']['id'][:]\t\t\t\t\t#not in the same order as gene_ids_xpr\n",
    "sample_ids_ctl = f['0']['META']['COL']['id'][:]\n",
    "matrix_ctl = pd.DataFrame(matrix_ctl, columns = sample_ids_ctl.astype(str), index = gene_ids_ctl.astype(int))\n",
    "del f\n",
    "\n",
    "#re-order gene_info based on the order in gene_ids_xpr (rows of data)\n",
    "gene_info.index = gene_info['gene_id']\n",
    "gene_info = gene_info.loc[gene_ids_oe.astype(int)].reset_index(inplace=False, drop=True)\n",
    "\n",
    "# output here is inst_info_xpr, inst_info_ctl, gene_info, matrix_xpr, matrix_ctl\n",
    "\n",
    "\n",
    "### STEP 2 FILTER METADATA\n",
    "#CRISPR\n",
    "list_ids = list(inst_info_oe['sample_id'])\t#in metadata\n",
    "matrix_oe = matrix_oe[list_ids]\t#Filtered data matrix\n",
    "#CONTROL\n",
    "list_ids = list(inst_info_ctl['sample_id'])\t#in metadata\n",
    "list_ids = list(set(list_ids).intersection(set(matrix_ctl.columns.astype(str))))\t#in metadata and in data matrix (some of metadata are not in data matrix)\n",
    "inst_info_ctl.index = inst_info_ctl['sample_id']; inst_info_ctl = inst_info_ctl.loc[list_ids].reset_index(inplace=False, drop=True) #remove entries from metadata that are not in data matrix\n",
    "matrix_ctl = matrix_ctl[list_ids]\t#Filtered data matrix\n",
    "# output here is inst_info_xpr, matrix_xpr, inst_info_ctl, matrix_ctl\n",
    "# by the end of this, inst_info_xpr is now # perturbed cells as N rows, matrix_xpr is # genes expressed X # perturbed cells. visa versa for ctl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cell_iname\n",
      "A549       17196\n",
      "PC3        12374\n",
      "A375       12001\n",
      "HEK293T    10647\n",
      "HA1E       10173\n",
      "MCF7        8453\n",
      "HT29        8104\n",
      "VCAP        5099\n",
      "HCC515      3814\n",
      "AALE        3164\n",
      "H1299       2858\n",
      "SALE        2445\n",
      "HEPG2       1291\n",
      "U2OS         561\n",
      "HUH7         314\n",
      "RWPE1        190\n",
      "LHSAR        180\n",
      "LNCAP        135\n",
      "HEK293        20\n",
      "HELA          16\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# print(inst_info_oe[['cell_iname', 'cmap_name']])\n",
    "# a list of unique cell lines found in oe\n",
    "# unique_cell_lines = set(inst_info_oe['cell_iname'])\n",
    "print(inst_info_oe[['cell_iname', 'cmap_name']].groupby(['cell_iname']).size().sort_values(ascending=False))\n",
    "\n",
    "# top cell lines that have > 4000 genes overexpressed:\n",
    "# A549, PC3, A375, HEK293T, HA1E, MCF7, HT29, VCAP\n",
    "\n",
    "# overlap of kept cell lines between xpr and oe is \n",
    "# A549, PC3, A375, MCF7, HT29"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# STEP 3 FILTER BY CELLLINE\n",
    "\n",
    "#####OE\n",
    "#Obtain cell lines with the most perturbations (> 4K genes perturbed)\n",
    "keep_cell_lines = ['A549', 'PC3', 'A375', 'HEK293T', 'HA1E', 'MCF7', 'HT29', 'VCAP']\n",
    "#Find indices of samples that are on the desired cell lines\n",
    "keep_index = []\n",
    "for i in range(len(inst_info_oe)):\n",
    "    if inst_info_oe.at[i, 'cell_iname'] in keep_cell_lines:\n",
    "        keep_index.append(i)\n",
    "inst_info_oe = inst_info_oe.loc[keep_index].reset_index(inplace=False, drop=True) #filter from metadata\n",
    "list_ids = list(inst_info_oe['sample_id'])\t#obtain sample ID from metadata\n",
    "matrix_oe = matrix_oe[list_ids]\t#Filtered data matrix\n",
    "#####CONTROL\n",
    "keep_index = []\n",
    "for i in range(len(inst_info_ctl)):\n",
    "    if inst_info_ctl.at[i, 'cell_iname'] in keep_cell_lines:\n",
    "        keep_index.append(i)\n",
    "inst_info_ctl = inst_info_ctl.loc[keep_index].reset_index(inplace=False, drop=True)  #filter from metadata\n",
    "list_ids = list(inst_info_ctl['sample_id'])\t#obtain sample ID from metadata\n",
    "matrix_ctl = matrix_ctl[list_ids]\t#Filtered data matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12328, 32018)\n",
      "(12328, 84047)\n",
      "(32018, 30)\n",
      "(84047, 30)\n"
     ]
    }
   ],
   "source": [
    "print(matrix_ctl.shape)\n",
    "print(matrix_oe.shape)\n",
    "\n",
    "print(inst_info_ctl.shape)\n",
    "print(inst_info_oe.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# STEO 4 FILTER UNKNOWN PERTURBED GENES (indeed unique to xpr, or the treatment group)\n",
    "# this matches the perturbed gene of each of the cells kept from previous steps to the gene names recorded in gene_info, if it's not there then drop\n",
    "known_genes = list(set(gene_info['gene_symbol']))\n",
    "keep_index = []\n",
    "for i in range(len(inst_info_oe)):\n",
    "    if inst_info_oe.at[i, 'cmap_name'] in known_genes:\n",
    "        keep_index.append(i)\n",
    "\n",
    "inst_info_oe = inst_info_oe.loc[keep_index].reset_index(inplace=False, drop=True) #filter from metadata\n",
    "list_ids = list(inst_info_oe['sample_id'])\t#obtain sample ID from metadata\n",
    "matrix_oe = matrix_oe[list_ids]\n",
    "# only inst_info_xpr, matrix_xpr are returned here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12328, 32018)\n",
      "(12328, 65010)\n",
      "(32018, 30)\n",
      "(65010, 30)\n"
     ]
    }
   ],
   "source": [
    "print(matrix_ctl.shape)\n",
    "print(matrix_oe.shape)\n",
    "\n",
    "print(inst_info_ctl.shape)\n",
    "print(inst_info_oe.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA7gAAAF1CAYAAADC0EYYAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/OQEPoAAAACXBIWXMAAAsTAAALEwEAmpwYAAAhU0lEQVR4nO3df9Sl5Vkf+u8lE1NiAoEwSXEGHZpgLdBKGg7BZrVHiwU0VrCLLCdHzVhpcaWJRz329ID9gSYLC+vEcJKlwaJQSEwDFLXhmFCcEl3WFgmTNIZfyWEUDBOmMM0AIW2DDrnOH/sZs2d4532HmQl73pvPZ6299rOv/dzPvjbPmmG+7/0891vdHQAAAFjtvm7RDQAAAMChIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAF4Qaqqe6vqOxbdxyJV1fdX1cNV9aWqeu0hPO6PVNXvH6rjAcD+EnABGE5VPVRV37VXbY/Q1d2ndPfvrnCcDVXVVbXma9Tqor0rydu7+6Xd/V8W3QwAHCwBFwAW5DAIzt+c5N4F9wAAh4yAC8AL0vwsb1WdUVVbquqLVfVoVb172u33pucnpst4v72qvq6q/llV/UlVPVZV76+qo+eO+5bpvS9U1T/f63N+tqpurqpfq6ovJvmR6bPvqKonqmp7Vf1iVX393PG6qv5RVT1QVU9V1Tur6tXTmC9W1U3z++/1HZfstapeXFVfSnJEkj+sqj9aYuwvV9W79qp9uKr+j2n74qr6o6mn+6rq+/fRw7Nmwavqd6vqH8y9/tGqur+qHq+q26rqm6d6VdWVU+9PVtWnq+rUfZ5UAF7wBFwASN6T5D3dfVSSVye5aar/ren55dNlvHck+ZHp8Z1J/lKSlyb5xSSpqpOTvC/JDyY5PsnRSdbt9VnnJbk5ycuTfDDJM0l+KslxSb49yVlJ/tFeY85N8rokZyb5J0munj7jhCSnJnnzPr7Xkr1299Pd/dJpn2/r7lcvMfbfJPmBqqrpux2T5OwkN0zv/1GSvzl9x59L8mtVdfw++tinqjo/yc8k+XtJ1ib5j0k+NL19dmbn4Fsy++/1A0m+8Fw/A4AXDgEXgFH9u2lW9ImqeiKz4Lkvf5bkNVV1XHd/qbv/YJl9fzDJu7v7j7v7S0kuSbJxmqG8IMn/292/391/muRfJOm9xt/R3f+uu7/S3f+zuz/R3X/Q3bu6+6Ek/yrJ/7rXmCu6+4vdfW+Se5L89vT5Tya5Ncm+FoharteV/Mep9785vb5g6v2RJOnuf9vdj0zf48YkDyQ5Yz+Ou7cfS/Ivu/v+7t6V5OeTnDbN4v5Zkpcl+dYkNe2z/QA+A4AXCAEXgFGd390v3/3Is2dF512Y2SzhZ6rqrqr63mX2/cYkfzL3+k+SrEnyqum9h3e/0d3/I8+ecXx4/kVVfUtV/VZV/dfpsuWfz2w2d96jc9v/c4nXL83Slut1Wd3dmc3W7p4d/t8ym3He3fdbqupTcz9AOHWJvvfHNyd5z9xxdiapJOu6+2OZzY7/UpJHq+rqqjrqAD4DgBcIAReAF7zufqC735zklUmuSHJzVX1Dnj37miSPZBbKdvumJLsyC53bk6zf/UZVHZnkFXt/3F6vr0rymSQnTZdI/0xmAe9QWK7X/fGhJBdMs6mvT/LrSTK9/pUkb0/yiukHCPdk6b7/+/T8krnaX5zbfjjJj83/MKK7j+zu/5wk3f3e7n5dklMy+yHE/7mfvQPwAiTgAvCCV1U/VFVru/srSZ6Yys8k2ZHkK5ndv7rbh5L8VFWdWFUvzWzG9cbp8tqbk/zdqvob08JPP5eVw+rLknwxyZeq6luTvPVQfa8Vel3R9KuDdiT51SS3dfcT01u7w/+OJKmqv5/ZDO5Sx9iR5PNJfqiqjqiqH83sPufdfjnJJVV1ynSso6vqTdP2/1JVr6+qF2UWlL+c2XkBgCUJuAAwW8Tp3mll4fck2djdX54uMb4syX+aLqE9M8m1ST6Q2QrLD2YWun48SaZ7ZH88s0t7tyd5KsljSZ5e5rP/cWaX/z6V2azojYfwe+2z1+fgQ0m+K7NFp5Ik3X1fkl9Ickdms8F/Ncl/WuYY/zCzmdcvZDYT+5/njvWbmc2a3zBdon1Pku+e3j4qs/8mj2d2efUXMvvdvQCwpJrdYgMAHGrTrOkTmV1+/OCC2wGA4ZnBBYBDqKr+blW9ZLqH911J7k7y0GK7AoAXBgEXAA6t8zJb3OmRJCdldrmzy6UA4HngEmUAAACGYAYXAACAIQi4AAAADGHNohs41I477rjesGHDotsAAADga+ATn/jEf+vutUu9N1zA3bBhQ7Zs2bLoNgAAAPgaqKo/2dd7LlEGAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIawZtENAKxmGy7+yKJbWPUeuvyNi24BABiEGVwAAACGYAYXgIUyC37wzIIDwIwZXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIawYsCtqr9QVR+vqj+sqnur6uem+rFVtbmqHpiej5kbc0lVba2qz1bVOXP111XV3dN7762qmuovrqobp/qdVbVhbsym6TMeqKpNh/TbAwAAMIz9mcF9Osnf7u5vS3JaknOr6swkFye5vbtPSnL79DpVdXKSjUlOSXJukvdV1RHTsa5KclGSk6bHuVP9wiSPd/drklyZ5IrpWMcmuTTJ65OckeTS+SANAAAAu60YcHvmS9PLF02PTnJekuun+vVJzp+2z0tyQ3c/3d0PJtma5IyqOj7JUd19R3d3kvfvNWb3sW5OctY0u3tOks3dvbO7H0+yOV8NxQAAAPDn9use3Ko6oqo+leSxzALnnUle1d3bk2R6fuW0+7okD88N3zbV1k3be9f3GNPdu5I8meQVyxwLAAAA9rBfAbe7n+nu05Ksz2w29tRldq+lDrFM/UDHfPUDqy6qqi1VtWXHjh3LtAYAAMContMqyt39RJLfzewy4Ueny44zPT827bYtyQlzw9YneWSqr1+ivseYqlqT5OgkO5c51t59Xd3dp3f36WvXrn0uXwkAAIBB7M8qymur6uXT9pFJvivJZ5LckmT3qsabknx42r4lycZpZeQTM1tM6uPTZcxPVdWZ0/21b9lrzO5jXZDkY9N9urclObuqjpkWlzp7qgEAAMAe1uzHPscnuX5aCfnrktzU3b9VVXckuamqLkzyuSRvSpLuvreqbkpyX5JdSd7W3c9Mx3prkuuSHJnk1umRJNck+UBVbc1s5nbjdKydVfXOJHdN+72ju3cezBcGAABgTCsG3O7+dJLXLlH/QpKz9jHmsiSXLVHfkuRZ9+9295czBeQl3rs2ybUr9QkAAMAL23O6BxcAAAAOVwIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCCsG3Ko6oap+p6rur6p7q+onpvrPVtXnq+pT0+N75sZcUlVbq+qzVXXOXP11VXX39N57q6qm+our6sapfmdVbZgbs6mqHpgemw7ptwcAAGAYa/Zjn11Jfrq7P1lVL0vyiaraPL13ZXe/a37nqjo5ycYkpyT5xiT/oaq+pbufSXJVkouS/EGSjyY5N8mtSS5M8nh3v6aqNia5IskPVNWxSS5NcnqSnj77lu5+/OC+NgAAAKNZcQa3u7d39yen7aeS3J9k3TJDzktyQ3c/3d0PJtma5IyqOj7JUd19R3d3kvcnOX9uzPXT9s1Jzppmd89Jsrm7d06hdnNmoRgAAAD28JzuwZ0uHX5tkjun0tur6tNVdW1VHTPV1iV5eG7Ytqm2btreu77HmO7eleTJJK9Y5lh793VRVW2pqi07dux4Ll8JAACAQex3wK2qlyb59SQ/2d1fzOxy41cnOS3J9iS/sHvXJYb3MvUDHfPVQvfV3X16d5++du3a5b4GAAAAg9qvgFtVL8os3H6wu38jSbr70e5+pru/kuRXkpwx7b4tyQlzw9cneWSqr1+ivseYqlqT5OgkO5c5FgAAAOxhf1ZRriTXJLm/u989Vz9+brfvT3LPtH1Lko3TysgnJjkpyce7e3uSp6rqzOmYb0ny4bkxu1dIviDJx6b7dG9LcnZVHTNdAn32VAMAAIA97M8qym9I8sNJ7q6qT021n0ny5qo6LbNLhh9K8mNJ0t33VtVNSe7LbAXmt00rKCfJW5Ncl+TIzFZPvnWqX5PkA1W1NbOZ243TsXZW1TuT3DXt947u3nkgXxQAAICxrRhwu/v3s/S9sB9dZsxlSS5bor4lyalL1L+c5E37ONa1Sa5dqU8AAABe2J7TKsoAAABwuBJwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQ1iz6AYAgIOz4eKPLLqFVe+hy9+46BYAOATM4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGMKKAbeqTqiq36mq+6vq3qr6ial+bFVtrqoHpudj5sZcUlVbq+qzVXXOXP11VXX39N57q6qm+our6sapfmdVbZgbs2n6jAeqatMh/fYAAAAMY39mcHcl+enu/itJzkzytqo6OcnFSW7v7pOS3D69zvTexiSnJDk3yfuq6ojpWFcluSjJSdPj3Kl+YZLHu/s1Sa5McsV0rGOTXJrk9UnOSHLpfJAGAACA3VYMuN29vbs/OW0/leT+JOuSnJfk+mm365OcP22fl+SG7n66ux9MsjXJGVV1fJKjuvuO7u4k799rzO5j3ZzkrGl295wkm7t7Z3c/nmRzvhqKAQAA4M89p3twp0uHX5vkziSv6u7tySwEJ3nltNu6JA/PDds21dZN23vX9xjT3buSPJnkFcsca+++LqqqLVW1ZceOHc/lKwEAADCI/Q64VfXSJL+e5Ce7+4vL7bpErZepH+iYrxa6r+7u07v79LVr1y7TGgAAAKPar4BbVS/KLNx+sLt/Yyo/Ol12nOn5sam+LckJc8PXJ3lkqq9for7HmKpak+ToJDuXORYAAADsYX9WUa4k1yS5v7vfPffWLUl2r2q8KcmH5+obp5WRT8xsMamPT5cxP1VVZ07HfMteY3Yf64IkH5vu070tydlVdcy0uNTZUw0AAAD2sGY/9nlDkh9OcndVfWqq/UySy5PcVFUXJvlckjclSXffW1U3JbkvsxWY39bdz0zj3prkuiRHJrl1eiSzAP2Bqtqa2cztxulYO6vqnUnumvZ7R3fvPLCvCgAAwMhWDLjd/ftZ+l7YJDlrH2MuS3LZEvUtSU5dov7lTAF5ifeuTXLtSn0CAADwwvacVlEGAACAw5WACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQ1gx4FbVtVX1WFXdM1f72ar6fFV9anp8z9x7l1TV1qr6bFWdM1d/XVXdPb333qqqqf7iqrpxqt9ZVRvmxmyqqgemx6ZD9q0BAAAYzv7M4F6X5Nwl6ld292nT46NJUlUnJ9mY5JRpzPuq6ohp/6uSXJTkpOmx+5gXJnm8u1+T5MokV0zHOjbJpUlen+SMJJdW1THP+RsCAADwgrBiwO3u30uycz+Pd16SG7r76e5+MMnWJGdU1fFJjuruO7q7k7w/yflzY66ftm9OctY0u3tOks3dvbO7H0+yOUsHbQAAADioe3DfXlWfni5h3j2zui7Jw3P7bJtq66btvet7jOnuXUmeTPKKZY71LFV1UVVtqaotO3bsOIivBAAAwGp1oAH3qiSvTnJaku1JfmGq1xL79jL1Ax2zZ7H76u4+vbtPX7t27TJtAwAAMKoDCrjd/Wh3P9PdX0nyK5ndI5vMZllPmNt1fZJHpvr6Jep7jKmqNUmOzuyS6H0dCwAAAJ7lgALudE/tbt+fZPcKy7ck2TitjHxiZotJfby7tyd5qqrOnO6vfUuSD8+N2b1C8gVJPjbdp3tbkrOr6pjpEuizpxoAAAA8y5qVdqiqDyX5jiTHVdW2zFY2/o6qOi2zS4YfSvJjSdLd91bVTUnuS7Irydu6+5npUG/NbEXmI5PcOj2S5JokH6iqrZnN3G6cjrWzqt6Z5K5pv3d09/4udgUAAMALzIoBt7vfvET5mmX2vyzJZUvUtyQ5dYn6l5O8aR/HujbJtSv1CAAAAAezijIAAAAcNgRcAAAAhiDgAgAAMIQV78EFABjdhos/sugWVr2HLn/jolsAMIMLAADAGARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCGsGHCr6tqqeqyq7pmrHVtVm6vqgen5mLn3LqmqrVX12ao6Z67+uqq6e3rvvVVVU/3FVXXjVL+zqjbMjdk0fcYDVbXpkH1rAAAAhrM/M7jXJTl3r9rFSW7v7pOS3D69TlWdnGRjklOmMe+rqiOmMVcluSjJSdNj9zEvTPJ4d78myZVJrpiOdWySS5O8PskZSS6dD9IAAAAwb8WA292/l2TnXuXzklw/bV+f5Py5+g3d/XR3P5hka5Izqur4JEd19x3d3Unev9eY3ce6OclZ0+zuOUk2d/fO7n48yeY8O2gDAABAkgO/B/dV3b09SabnV071dUkenttv21RbN23vXd9jTHfvSvJkklcsc6xnqaqLqmpLVW3ZsWPHAX4lAAAAVrNDvchULVHrZeoHOmbPYvfV3X16d5++du3a/WoUAACAsRxowH10uuw40/NjU31bkhPm9luf5JGpvn6J+h5jqmpNkqMzuyR6X8cCAACAZznQgHtLkt2rGm9K8uG5+sZpZeQTM1tM6uPTZcxPVdWZ0/21b9lrzO5jXZDkY9N9urclObuqjpkWlzp7qgEAAMCzrFlph6r6UJLvSHJcVW3LbGXjy5PcVFUXJvlckjclSXffW1U3Jbkvya4kb+vuZ6ZDvTWzFZmPTHLr9EiSa5J8oKq2ZjZzu3E61s6qemeSu6b93tHdey92BQAAAEn2I+B295v38dZZ+9j/siSXLVHfkuTUJepfzhSQl3jv2iTXrtQjAAAAHOpFpgAAAGAhBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAhrFt0AsBgbLv7IolsAYCD+v3LwHrr8jYtuAVY9M7gAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAM4aACblU9VFV3V9WnqmrLVDu2qjZX1QPT8zFz+19SVVur6rNVdc5c/XXTcbZW1Xurqqb6i6vqxql+Z1VtOJh+AQAAGNehmMH9zu4+rbtPn15fnOT27j4pye3T61TVyUk2JjklyblJ3ldVR0xjrkpyUZKTpse5U/3CJI9392uSXJnkikPQLwAAAAP6WlyifF6S66ft65OcP1e/obuf7u4Hk2xNckZVHZ/kqO6+o7s7yfv3GrP7WDcnOWv37C4AAADMO9iA20l+u6o+UVUXTbVXdff2JJmeXznV1yV5eG7stqm2btreu77HmO7eleTJJK84yJ4BAAAY0JqDHP+G7n6kql6ZZHNVfWaZfZeaee1l6suN2fPAs3B9UZJ80zd90/IdAwAAMKSDmsHt7kem58eS/GaSM5I8Ol12nOn5sWn3bUlOmBu+PskjU339EvU9xlTVmiRHJ9m5RB9Xd/fp3X362rVrD+YrAQAAsEodcMCtqm+oqpft3k5ydpJ7ktySZNO026YkH562b0mycVoZ+cTMFpP6+HQZ81NVdeZ0f+1b9hqz+1gXJPnYdJ8uAAAA7OFgLlF+VZLfnNZ8WpPk33T3v6+qu5LcVFUXJvlckjclSXffW1U3Jbkvya4kb+vuZ6ZjvTXJdUmOTHLr9EiSa5J8oKq2ZjZzu/Eg+gUAAGBgBxxwu/uPk3zbEvUvJDlrH2MuS3LZEvUtSU5dov7lTAEZAAAAlvO1+DVBAAAA8LwTcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADGHNohsAAACSDRd/ZNEtDOGhy9+46BZYIDO4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIaxZdANwIDZc/JFFtwAAABxmBFwAAGAYJkIO3kOXv3HRLRwwlygDAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABjCqvg1QVV1bpL3JDkiya929+ULbumgWLocAADg0DvsZ3Cr6ogkv5Tku5OcnOTNVXXyYrsCAADgcHPYB9wkZyTZ2t1/3N1/muSGJOctuCcAAAAOM6sh4K5L8vDc621TDQAAAP7cargHt5ao9R47VF2U5KLp5Zeq6rNf864OznFJ/tuim2C/OV+ri/O1+jhnq4vztbo4X6uL87W6DHu+6opFd7Cib97XG6sh4G5LcsLc6/VJHpnfobuvTnL189nUwaiqLd19+qL7YP84X6uL87X6OGeri/O1ujhfq4vztbo4X4en1XCJ8l1JTqqqE6vq65NsTHLLgnsCAADgMHPYz+B2966qenuS2zL7NUHXdve9C24LAACAw8xhH3CTpLs/muSji+7jEFo1l1OTxPlabZyv1cc5W12cr9XF+VpdnK/Vxfk6DFV3r7wXAAAAHOZWwz24AAAAsCIB93lUVedW1WeramtVXbzoflheVZ1QVb9TVfdX1b1V9ROL7omVVdURVfVfquq3Ft0Ly6uql1fVzVX1menP2bcvuif2rap+avq78J6q+lBV/YVF98Sequraqnqsqu6Zqx1bVZur6oHp+ZhF9shX7eN8/d/T34mfrqrfrKqXL7BF5ix1vube+8dV1VV13CJ6Y08C7vOkqo5I8ktJvjvJyUneXFUnL7YrVrAryU93919JcmaStzlnq8JPJLl/0U2wX96T5N9397cm+bY4b4etqlqX5H9Pcnp3n5rZoo8bF9sVS7guybl71S5Ocnt3n5Tk9uk1h4fr8uzztTnJqd3915L8f0kueb6bYp+uy7PPV6rqhCR/J8nnnu+GWJqA+/w5I8nW7v7j7v7TJDckOW/BPbGM7t7e3Z+ctp/K7B/f6xbbFcupqvVJ3pjkVxfdC8urqqOS/K0k1yRJd/9pdz+x0KZYyZokR1bVmiQvyV6/k57F6+7fS7Jzr/J5Sa6ftq9Pcv7z2RP7ttT56u7f7u5d08s/SLL+eW+MJe3jz1eSXJnknySxsNFhQsB9/qxL8vDc620RllaNqtqQ5LVJ7lxwKyzv/8nsfzJfWXAfrOwvJdmR5F9Pl5T/alV9w6KbYmnd/fkk78pshmJ7kie7+7cX2xX76VXdvT2Z/eA2ySsX3A/770eT3LroJti3qvq+JJ/v7j9cdC98lYD7/Kklan7SswpU1UuT/HqSn+zuLy66H5ZWVd+b5LHu/sSie2G/rEny15Nc1d2vTfLf49LJw9Z03+Z5SU5M8o1JvqGqfmixXcG4quqfZnar1AcX3QtLq6qXJPmnSf7FonthTwLu82dbkhPmXq+Py7sOe1X1oszC7Qe7+zcW3Q/LekOS76uqhzK7BeBvV9WvLbYllrEtybbu3n1VxM2ZBV4OT9+V5MHu3tHdf5bkN5L8jQX3xP55tKqOT5Lp+bEF98MKqmpTku9N8oPt93kezl6d2Q/9/nD6t8f6JJ+sqr+40K4QcJ9HdyU5qapOrKqvz2xxjlsW3BPLqKrK7P7A+7v73Yvuh+V19yXdvb67N2T25+tj3W2G6TDV3f81ycNV9Zen0llJ7ltgSyzvc0nOrKqXTH83nhWLgq0WtyTZNG1vSvLhBfbCCqrq3CT/V5Lv6+7/seh+2Lfuvru7X9ndG6Z/e2xL8ten/7+xQALu82RaMODtSW7L7B8FN3X3vYvtihW8IckPZzYT+Knp8T2LbgoG8uNJPlhVn05yWpKfX2w77Ms0035zkk8muTuzfz9cvdCmeJaq+lCSO5L85araVlUXJrk8yd+pqgcyW+n18kX2yFft43z9YpKXJdk8/bvjlxfaJH9uH+eLw1C58gEAAIARmMEFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAM4f8HqypulTH+0noAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1152x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# STEP 5 NORMALIZE AND SAVE TO NPZ\n",
    "# note all previous variables were passed here and this function simply returned as it normalizes and saves to a directory. \n",
    "\n",
    "# note here they concatenated the ctl and the xpr indices because indeed they wouldn't overlap as the two lists were taken from the inst info in the first place\n",
    "metadata = pd.concat([inst_info_oe, inst_info_ctl], axis=0).reset_index(inplace=False, drop=True)\n",
    "# here they create the file that records each cell id and their metadata attributes\n",
    "# metadata.to_csv(osp.join(outdir, 'all_metadata.txt'))\n",
    "# !!!!!! here they concatenated the gene matrices, by extending the cells dimension (gene dimension stays the same)\n",
    "matrix = pd.concat([matrix_oe, matrix_ctl], axis=1)\n",
    "del(matrix_oe)\n",
    "\n",
    "# note the below histogram shows that most values (of the .1% of all samples in the concatenated matrix) have a value of ~6\n",
    "mv = matrix.values.flatten() # collapses value into one line for sampling\n",
    "sampling = sample(range(len(mv)), int(0.001*len(mv)))  # pick 1/1000 of the total samples in the big matrix\n",
    "mv = mv[sampling]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16,6))\n",
    "ax.hist(mv)\n",
    "ax.set_title('Histogram of values')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA7gAAAF1CAYAAADC0EYYAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/OQEPoAAAACXBIWXMAAAsTAAALEwEAmpwYAAAgxklEQVR4nO3dfbSl1V0f8O9PJqZoAuFlkuKAGZpgFVKDZSRYa43iAhJrSbrIcuILqLTYmLjU2lZwtaLJwoa1NKmphhQNKyRqgOJL0ATjNNHGF0KYWAwBpIwBwwiFMUMIUYMO+fWP84yeudy59zJvd2bP57PWWeec37P3PvvMPGtmfe9+nn2ruwMAAACHui9Y7QkAAADAviDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABOCxV1Z1V9dLVnsdqqqpXVtUDVfXZqvqqfTjud1XV7++r8QBgpQRcAIZTVfdX1TctqO0Surr7tO7+3WXGWV9VXVVr9tNUV9tPJXlddz+ru//Pak8GAPaWgAsAq+QgCM7PT3LnKs8BAPYZAReAw9L8Km9VnVlVm6vqM1X1cFW9aWr2oen509NlvF9TVV9QVf+5qv6sqh6pqndW1dFz4144HftUVf2XBZ/z41V1Y1X9YlV9Jsl3TZ99S1V9uqoeqqqfraovnBuvq+r7qureqnq8qt5QVS+Y+nymqm6Yb7/gOy4616p6ZlV9NskRSf64qv50kb5vq6qfWlB7T1X9++n1pVX1p9Oc7qqqV+5mDk9ZBa+q362qfzP3/nuq6u6qerSq3l9Vz5/qVVVvnub+WFV9rKpetNu/VAAOewIuACQ/k+RnuvuoJC9IcsNU/xfT83Omy3hvSfJd0+MbkvyjJM9K8rNJUlWnJnlrkm9PckKSo5OsW/BZ5ye5MclzkvxSkieT/FCS45N8TZKzk3zfgj7nJTkjyVlJ/lOSq6fPOCnJi5K8ejffa9G5dvcT3f2sqc2Lu/sFi/T95STfWlU1fbdjkpyT5Lrp+J8m+brpO/5Ekl+sqhN2M4/dqqpXJPnRJP86ydokv5fk3dPhczL7O/iyzP68vjXJp57uZwBw+BBwARjVr0+rop+uqk9nFjx352+TvLCqju/uz3b3h5do++1J3tTdn+juzya5LMnGaYXygiS/0d2/391/k+THkvSC/rd096939+e7+6+7+6Pd/eHu3tHd9yf5H0m+fkGfK7v7M919Z5KPJ/nt6fMfS3Jzkt1tELXUXJfze9Pcv256f8E09weTpLv/Z3c/OH2P65Pcm+TMFYy70Pcm+a/dfXd370jyk0lOn1Zx/zbJs5N8eZKa2jy0B58BwGFCwAVgVK/o7ufsfOSpq6LzLs5slfBPquq2qvqXS7T9kiR/Nvf+z5KsSfK86dgDOw9091/lqSuOD8y/qaovq6rfrKr/N122/JOZrebOe3ju9V8v8v5ZWdxSc11Sd3dmq7U7V4e/LbMV553zvrCqbp/7AcKLFpn3Sjw/yc/MjbM9SSVZ190fzGx1/OeSPFxVV1fVUXvwGQAcJgRcAA573X1vd786yXOTXJnkxqr64jx19TVJHswslO30pUl2ZBY6H0py4s4DVXVkkuMWftyC91cl+ZMkp0yXSP9oZgFvX1hqrivx7iQXTKupL0nyK0kyvf/5JK9Lctz0A4SPZ/F5/+X0/EVztX849/qBJN87/8OI7j6yu/8wSbr7Ld19RpLTMvshxH9c4dwBOAwJuAAc9qrqO6pqbXd/Psmnp/KTSbYl+Xxm96/u9O4kP1RVJ1fVszJbcb1+urz2xiTfUlX/bNr46SeyfFh9dpLPJPlsVX15ktfsq++1zFyXNf3qoG1JfiHJ+7v709OhneF/W5JU1XdntoK72Bjbkvx5ku+oqiOq6nsyu895p7cluayqTpvGOrqqXjW9/uqqeklVPSOzoPy5zP5eAGBRAi4AzDZxunPaWfhnkmzs7s9NlxhfkeQPpktoz0pyTZJ3ZbbD8n2Zha7vT5LpHtnvz+zS3oeSPJ7kkSRPLPHZ/yGzy38fz2xV9Pp9+L12O9en4d1JvimzTaeSJN19V5KfTnJLZqvB/yTJHywxxr/NbOX1U5mtxP7h3Fi/ltmq+XXTJdofT/Ky6fBRmf2ZPJrZ5dWfyux39wLAomp2iw0AsK9Nq6afzuzy4/tWeToAMDwruACwD1XVt1TVF0338P5UkjuS3L+6swKAw4OACwD71vmZbe70YJJTMrvc2eVSAHAAuEQZAACAIVjBBQAAYAgCLgAAAENYs9oT2NeOP/74Xr9+/WpPAwAAgP3gox/96F9099rFjg0XcNevX5/Nmzev9jQAAADYD6rqz3Z3zCXKAAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAENYNuBW1T+oqo9U1R9X1Z1V9RNT/diq2lRV907Px8z1uayqtlTVPVV17lz9jKq6Yzr2lqqqqf7Mqrp+qt9aVevn+lw0fca9VXXRPv32AAAADGMlK7hPJPnG7n5xktOTnFdVZyW5NMkHuvuUJB+Y3qeqTk2yMclpSc5L8taqOmIa66oklyQ5ZXqcN9UvTvJod78wyZuTXDmNdWySy5O8JMmZSS6fD9IAAACw05rlGnR3J/ns9PYZ06OTnJ/kpVP92iS/m+RHpvp13f1EkvuqakuSM6vq/iRHdfctSVJV70zyiiQ3T31+fBrrxiQ/O63unptkU3dvn/psyiwUv3sPvy8AwFOsv/S9qz2FQ979b/zm1Z4CwMruwa2qI6rq9iSPZBY4b03yvO5+KEmm5+dOzdcleWCu+9aptm56vbC+S5/u3pHksSTHLTEWAAAA7GJFAbe7n+zu05OcmNlq7IuWaF6LDbFEfU/7/P0HVl1SVZuravO2bduWmBoAAACjelq7KHf3pzO7FPm8JA9X1QlJMj0/MjXbmuSkuW4nJnlwqp+4SH2XPlW1JsnRSbYvMdbCeV3d3Ru6e8PatWufzlcCAABgECvZRXltVT1nen1kkm9K8idJbkqyc1fji5K8Z3p9U5KN087IJ2e2mdRHpsuYH6+qs6b7ay9c0GfnWBck+eB07+/7k5xTVcdMm0udM9UAAABgF8tuMpXkhCTXTjshf0GSG7r7N6vqliQ3VNXFST6Z5FVJ0t13VtUNSe5KsiPJa7v7yWms1yR5R5IjM9tc6uap/vYk75o2pNqe2S7M6e7tVfWGJLdN7V6/c8MpAAAAmLeSXZQ/luSrFql/KsnZu+lzRZIrFqlvTvKU+3e7+3OZAvIix65Jcs1y8wQAAODw9rTuwQUAAICDlYALAADAEARcAAAAhiDgAgAAMISV7KIMABzE1l/63tWeAgAcFKzgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYwrIBt6pOqqrfqaq7q+rOqvqBqf7jVfXnVXX79Hj5XJ/LqmpLVd1TVefO1c+oqjumY2+pqprqz6yq66f6rVW1fq7PRVV17/S4aJ9+ewAAAIaxZgVtdiT54e7+o6p6dpKPVtWm6dibu/un5htX1alJNiY5LcmXJPlfVfVl3f1kkquSXJLkw0nel+S8JDcnuTjJo939wqramOTKJN9aVccmuTzJhiQ9ffZN3f3o3n1tAAAARrPsCm53P9TdfzS9fjzJ3UnWLdHl/CTXdfcT3X1fki1JzqyqE5Ic1d23dHcneWeSV8z1uXZ6fWOSs6fV3XOTbOru7VOo3ZRZKAYAAIBdPK17cKdLh78qya1T6XVV9bGquqaqjplq65I8MNdt61RbN71eWN+lT3fvSPJYkuOWGGvhvC6pqs1VtXnbtm1P5ysBAAAwiBUH3Kp6VpJfSfKD3f2ZzC43fkGS05M8lOSndzZdpHsvUd/TPn9f6L66uzd094a1a9cu9TUAAAAY1IoCblU9I7Nw+0vd/atJ0t0Pd/eT3f35JD+f5Myp+dYkJ811PzHJg1P9xEXqu/SpqjVJjk6yfYmxAAAAYBcr2UW5krw9yd3d/aa5+glzzV6Z5OPT65uSbJx2Rj45ySlJPtLdDyV5vKrOmsa8MMl75vrs3CH5giQfnO7TfX+Sc6rqmOkS6HOmGgAAAOxiJbsof22S70xyR1XdPtV+NMmrq+r0zC4Zvj/J9yZJd99ZVTckuSuzHZhfO+2gnCSvSfKOJEdmtnvyzVP97UneVVVbMlu53TiNtb2q3pDktqnd67t7+558UQAAAMa2bMDt7t/P4vfCvm+JPlckuWKR+uYkL1qk/rkkr9rNWNckuWa5eQIAAHB4e1q7KAMAAMDBSsAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCMsG3Ko6qap+p6rurqo7q+oHpvqxVbWpqu6dno+Z63NZVW2pqnuq6ty5+hlVdcd07C1VVVP9mVV1/VS/tarWz/W5aPqMe6vqon367QEAABjGSlZwdyT54e7+iiRnJXltVZ2a5NIkH+juU5J8YHqf6djGJKclOS/JW6vqiGmsq5JckuSU6XHeVL84yaPd/cIkb05y5TTWsUkuT/KSJGcmuXw+SAMAAMBOywbc7n6ou/9oev14kruTrEtyfpJrp2bXJnnF9Pr8JNd19xPdfV+SLUnOrKoTkhzV3bd0dyd554I+O8e6McnZ0+ruuUk2dff27n40yab8fSgGAACAv/O07sGdLh3+qiS3Jnledz+UzEJwkudOzdYleWCu29aptm56vbC+S5/u3pHksSTHLTEWAAAA7GLFAbeqnpXkV5L8YHd/Zqmmi9R6ifqe9pmf2yVVtbmqNm/btm2JqQEAADCqFQXcqnpGZuH2l7r7V6fyw9Nlx5meH5nqW5OcNNf9xCQPTvUTF6nv0qeq1iQ5Osn2JcbaRXdf3d0bunvD2rVrV/KVAAAAGMya5RpM98K+Pcnd3f2muUM3JbkoyRun5/fM1X+5qt6U5Esy20zqI939ZFU9XlVnZXaJ84VJ/vuCsW5JckGSD3Z3V9X7k/zk3MZS5yS5bI+/LQAA+8X6S9+72lM45N3/xm9e7SnAIW/ZgJvka5N8Z5I7qur2qfajmQXbG6rq4iSfTPKqJOnuO6vqhiR3ZbYD82u7+8mp32uSvCPJkUlunh7JLEC/q6q2ZLZyu3Eaa3tVvSHJbVO713f39j37qgAAAIysZhsaj2PDhg29efPm1Z4GACtk1QdgxgourExVfbS7Nyx27GntogwAAAAHKwEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADEHABQAAYAgCLgAAAEMQcAEAABiCgAsAAMAQBFwAAACGIOACAAAwBAEXAACAIQi4AAAADGHZgFtV11TVI1X18bnaj1fVn1fV7dPj5XPHLquqLVV1T1WdO1c/o6rumI69papqqj+zqq6f6rdW1fq5PhdV1b3T46J99q0BAAAYzkpWcN+R5LxF6m/u7tOnx/uSpKpOTbIxyWlTn7dW1RFT+6uSXJLklOmxc8yLkzza3S9M8uYkV05jHZvk8iQvSXJmksur6pin/Q0BAAA4LCwbcLv7Q0m2r3C885Nc191PdPd9SbYkObOqTkhyVHff0t2d5J1JXjHX59rp9Y1Jzp5Wd89Nsqm7t3f3o0k2ZfGgDQAAAHt1D+7rqupj0yXMO1dW1yV5YK7N1qm2bnq9sL5Ln+7ekeSxJMctMdZTVNUlVbW5qjZv27ZtL74SAAAAh6o9DbhXJXlBktOTPJTkp6d6LdK2l6jvaZ9di91Xd/eG7t6wdu3aJaYNAADAqPYo4Hb3w939ZHd/PsnPZ3aPbDJbZT1prumJSR6c6icuUt+lT1WtSXJ0ZpdE724sAAAAeIo9CrjTPbU7vTLJzh2Wb0qycdoZ+eTMNpP6SHc/lOTxqjprur/2wiTvmeuzc4fkC5J8cLpP9/1JzqmqY6ZLoM+ZagAAAPAUa5ZrUFXvTvLSJMdX1dbMdjZ+aVWdntklw/cn+d4k6e47q+qGJHcl2ZHktd395DTUazLbkfnIJDdPjyR5e5J3VdWWzFZuN05jba+qNyS5bWr3+u5e6WZXAAAAHGaWDbjd/epFym9fov0VSa5YpL45yYsWqX8uyat2M9Y1Sa5Zbo4AAACwN7soAwAAwEFDwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIywbcqrqmqh6pqo/P1Y6tqk1Vde/0fMzcscuqaktV3VNV587Vz6iqO6Zjb6mqmurPrKrrp/qtVbV+rs9F02fcW1UX7bNvDQAAwHBWsoL7jiTnLahdmuQD3X1Kkg9M71NVpybZmOS0qc9bq+qIqc9VSS5Jcsr02DnmxUke7e4XJnlzkiunsY5NcnmSlyQ5M8nl80EaAAAA5i0bcLv7Q0m2Lyifn+Ta6fW1SV4xV7+uu5/o7vuSbElyZlWdkOSo7r6luzvJOxf02TnWjUnOnlZ3z02yqbu3d/ejSTblqUEbAAAAkuz5PbjP6+6HkmR6fu5UX5fkgbl2W6fauun1wvoufbp7R5LHkhy3xFgAAADwFPt6k6lapNZL1Pe0z64fWnVJVW2uqs3btm1b0UQBAAAYy54G3Ieny44zPT8y1bcmOWmu3YlJHpzqJy5S36VPVa1JcnRml0Tvbqyn6O6ru3tDd29Yu3btHn4lAAAADmV7GnBvSrJzV+OLkrxnrr5x2hn55Mw2k/rIdBnz41V11nR/7YUL+uwc64IkH5zu031/knOq6phpc6lzphoAAAA8xZrlGlTVu5O8NMnxVbU1s52N35jkhqq6OMknk7wqSbr7zqq6IcldSXYkeW13PzkN9ZrMdmQ+MsnN0yNJ3p7kXVW1JbOV243TWNur6g1Jbpvavb67F252BQAAAElWEHC7+9W7OXT2btpfkeSKReqbk7xokfrnMgXkRY5dk+Sa5eYIAAAA+3qTKQAAAFgVAi4AAABDEHABAAAYgoALAADAEJbdZAqA3Vt/6XtXewoAAEys4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMYc1qTwAAAEjWX/re1Z7CEO5/4zev9hRYRVZwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBAEXAAAAIYg4AIAADAEARcAAIAhCLgAAAAMQcAFAABgCAIuAAAAQxBwAQAAGIKACwAAwBD2KuBW1f1VdUdV3V5Vm6fasVW1qarunZ6PmWt/WVVtqap7qurcufoZ0zhbquotVVVT/ZlVdf1Uv7Wq1u/NfAEAABjXvljB/YbuPr27N0zvL03yge4+JckHpvepqlOTbExyWpLzkry1qo6Y+lyV5JIkp0yP86b6xUke7e4XJnlzkiv3wXwBAAAY0P64RPn8JNdOr69N8oq5+nXd/UR335dkS5Izq+qEJEd19y3d3UneuaDPzrFuTHL2ztVdAAAAmLe3AbeT/HZVfbSqLplqz+vuh5Jken7uVF+X5IG5vlun2rrp9cL6Ln26e0eSx5Ict3ASVXVJVW2uqs3btm3by68EAADAoWjNXvb/2u5+sKqem2RTVf3JEm0XW3ntJepL9dm10H11kquTZMOGDU85DgAAwPj2agW3ux+cnh9J8mtJzkzy8HTZcabnR6bmW5OcNNf9xCQPTvUTF6nv0qeq1iQ5Osn2vZkzAAAAY9rjgFtVX1xVz975Osk5ST6e5KYkF03NLkrynun1TUk2Tjsjn5zZZlIfmS5jfryqzprur71wQZ+dY12Q5IPTfboAAACwi725RPl5SX5t2vNpTZJf7u7fqqrbktxQVRcn+WSSVyVJd99ZVTckuSvJjiSv7e4np7Fek+QdSY5McvP0SJK3J3lXVW3JbOV2417MFwAAgIHtccDt7k8kefEi9U8lOXs3fa5IcsUi9c1JXrRI/XOZAjIAAAAsZX/8miAAAAA44ARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMIQ1qz0BYHWsv/S9qz0FAADYp6zgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxhzWpPAAAAYF9Zf+l7V3sKh7z73/jNqz2FPWYFFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEARcAAAAhiDgAgAAMAQBFwAAgCGsWe0JwJ5Yf+l7V3sKAADAQeaQWMGtqvOq6p6q2lJVl672fAAAADj4HPQBt6qOSPJzSV6W5NQkr66qU1d3VgAAABxsDoVLlM9MsqW7P5EkVXVdkvOT3LWqs9oLLq8FAADY9w76Fdwk65I8MPd+61QDAACAv3MorODWIrXepUHVJUkumd5+tqru2e+z2jvHJ/mL1Z4Ehz3nIQcL5yIHA+chBwPnIQeFuvKgPxefv7sDh0LA3ZrkpLn3JyZ5cL5Bd1+d5OoDOam9UVWbu3vDas+Dw5vzkIOFc5GDgfOQg4HzkIPFoXwuHgqXKN+W5JSqOrmqvjDJxiQ3rfKcAAAAOMgc9Cu43b2jql6X5P1JjkhyTXffucrTAgAA4CBz0AfcJOnu9yV532rPYx86ZC6nZmjOQw4WzkUOBs5DDgbOQw4Wh+y5WN29fCsAAAA4yB0K9+ACAADAsgTc/aiqzquqe6pqS1Vdusjxqqq3TMc/VlX/dDXmydhWcB5++3T+fayq/rCqXrwa82Rsy52Hc+2+uqqerKoLDuT8OHys5FysqpdW1e1VdWdV/e8DPUfGt4L/m4+uqt+oqj+ezsPvXo15MraquqaqHqmqj+/m+CGZVQTc/aSqjkjyc0leluTUJK+uqlMXNHtZklOmxyVJrjqgk2R4KzwP70vy9d39lUnekEP4ngsOTis8D3e2uzKzTQVhn1vJuVhVz0ny1iT/qrtPS/KqAz1PxrbCfxNfm+Su7n5xkpcm+enpt4nAvvSOJOctcfyQzCoC7v5zZpIt3f2J7v6bJNclOX9Bm/OTvLNnPpzkOVV1woGeKENb9jzs7j/s7kentx/O7HdNw760kn8Pk+T7k/xKkkcO5OQ4rKzkXPy2JL/a3Z9Mku52PrKvreQ87CTPrqpK8qwk25PsOLDTZHTd/aHMzq3dOSSzioC7/6xL8sDc+61T7em2gb3xdM+xi5PcvF9nxOFo2fOwqtYleWWStx3AeXH4Wcm/iV+W5Jiq+t2q+mhVXXjAZsfhYiXn4c8m+YokDya5I8kPdPfnD8z04O8cklnlkPg1QYeoWqS2cMvqlbSBvbHic6yqviGzgPvP9+uMOByt5Dz8b0l+pLufnC1YwH6xknNxTZIzkpyd5Mgkt1TVh7v7/+7vyXHYWMl5eG6S25N8Y5IXJNlUVb/X3Z/Zz3ODeYdkVhFw95+tSU6ae39iZj+Fe7ptYG+s6Byrqq9M8gtJXtbdnzpAc+PwsZLzcEOS66Zwe3ySl1fVju7+9QMyQw4XK/2/+S+6+y+T/GVVfSjJi5MIuOwrKzkPvzvJG3v2+zy3VNV9Sb48yUcOzBQhySGaVVyivP/cluSUqjp52hRgY5KbFrS5KcmF0w5lZyV5rLsfOtATZWjLnodV9aVJfjXJd1qhYD9Z9jzs7pO7e313r09yY5LvE27ZD1byf/N7knxdVa2pqi9K8pIkdx/geTK2lZyHn8zsKoJU1fOS/OMknzigs4RDNKtYwd1PuntHVb0us91Aj0hyTXffWVX/bjr+tiTvS/LyJFuS/FVmP62DfWaF5+GPJTkuyVun1bMd3b1htebMeFZ4HsJ+t5JzsbvvrqrfSvKxJJ9P8gvdveiv0IA9scJ/E9+Q5B1VdUdml4n+SHf/xapNmiFV1bsz26X7+KramuTyJM9IDu2sUrMrHwAAAODQ5hJlAAAAhiDgAgAAMAQBFwAAgCEIuAAAAAxBwAUAAGAIAi4AAABDEHABAAAYgoALAADAEP4/fJCkSZfVtSMAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1152x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#NORMALIZATION\n",
    "# log + min max normalize, after this the values of matrix will be between 0 and 1\n",
    "matrix = np.log2(matrix + 1)\n",
    "scaler = MinMaxScaler((0,1))\n",
    "matrix = matrix.transpose()\n",
    "matrix = pd.DataFrame(scaler.fit_transform(matrix), columns = matrix.columns, index = matrix.index)\n",
    "matrix = matrix.transpose()\n",
    "\n",
    "#hist of values after lognorm\n",
    "mv = matrix.values.flatten()\n",
    "sampling = sample(range(len(mv)), int(0.001*len(mv)))\n",
    "mv = mv[sampling]\n",
    "\n",
    "# this gives that most values are at 0.6, and this is a very nice normal distribution \n",
    "fig, ax = plt.subplots(figsize=(16,6))\n",
    "ax.hist(mv)\n",
    "ax.set_title('Histogram of values')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2. Save data and metadata for each condition and cell line\n",
    "#Compound + cell lines\n",
    "#Control + cell lines\n",
    "outdir = \"/home/b-evelyntong/hl/lincs_lvl3_oe/processed\"\n",
    "metadata.index = metadata['sample_id']\n",
    "metadata = metadata.loc[matrix.columns]\t#sort metadata given by column order in data matrix (and filter samples that have been filtered out from matrix during binarization)\n",
    "for cell_line in keep_cell_lines:\n",
    "    for pert_type in ['trt_oe', 'ctl_vector']:\n",
    "        metadata_i = metadata[np.logical_and(metadata['cell_iname'] == cell_line, metadata['pert_type'] == pert_type)]\n",
    "        data_i = matrix[metadata_i.index]\n",
    "        metadata_i.to_csv(osp.join(outdir, 'cell_line_{}_pert_{}_metadata.txt'.format(cell_line, pert_type)), index=False)\n",
    "        filename = 'cell_line_{}_pert_{}'.format(cell_line, pert_type)\n",
    "        np.savez_compressed(osp.join(outdir, filename), data=data_i.values, row_ids = data_i.index, col_ids=data_i.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_iname</th>\n",
       "      <th>cellosaurus_id</th>\n",
       "      <th>donor_age</th>\n",
       "      <th>donor_age_death</th>\n",
       "      <th>donor_disease_age_onset</th>\n",
       "      <th>doubling_time</th>\n",
       "      <th>growth_medium</th>\n",
       "      <th>provider_catalog_id</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>donor_ethnicity</th>\n",
       "      <th>donor_sex</th>\n",
       "      <th>donor_tumor_phase</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>subtype</th>\n",
       "      <th>provider_name</th>\n",
       "      <th>growth_pattern</th>\n",
       "      <th>ccle_name</th>\n",
       "      <th>cell_alias</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HEK293T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>293T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>HA1E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60</td>\n",
       "      <td>MEM-ALPHA (Invitrogen A1049001) supplemented w...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>kidney</td>\n",
       "      <td>normal kidney sample</td>\n",
       "      <td>normal kidney sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HA1E_KIDNEY</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>A549</td>\n",
       "      <td>CVCL_0023</td>\n",
       "      <td>58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48</td>\n",
       "      <td>F-12K ATCC catalog # 3-24</td>\n",
       "      <td>CCL-185</td>\n",
       "      <td>c-4</td>\n",
       "      <td>tumor</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>M</td>\n",
       "      <td>Primary</td>\n",
       "      <td>lung</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>non small cell carcinoma</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>adherent</td>\n",
       "      <td>A549_LUNG</td>\n",
       "      <td>A 549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>A375</td>\n",
       "      <td>CVCL_0132</td>\n",
       "      <td>54</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36</td>\n",
       "      <td>DMEM Invitrogen catalog # 11995-65</td>\n",
       "      <td>CRL-1619</td>\n",
       "      <td>c-127</td>\n",
       "      <td>tumor</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>F</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin cancer</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>adherent</td>\n",
       "      <td>A375_SKIN</td>\n",
       "      <td>A 375|A-375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>HT29</td>\n",
       "      <td>CVCL_0320</td>\n",
       "      <td>44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36</td>\n",
       "      <td>McCoy's 5A Invitrogen catalog # 166-82</td>\n",
       "      <td>HTB-38</td>\n",
       "      <td>c-272</td>\n",
       "      <td>tumor</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>F</td>\n",
       "      <td>Primary</td>\n",
       "      <td>large_intestine</td>\n",
       "      <td>colon cancer</td>\n",
       "      <td>adenocarcinoma</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>adherent</td>\n",
       "      <td>HT29_LARGE_INTESTINE</td>\n",
       "      <td>HT 29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>MCF7</td>\n",
       "      <td>CVCL_0031</td>\n",
       "      <td>40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72</td>\n",
       "      <td>EMEM ATCC catalog # 3-23</td>\n",
       "      <td>HTB-22</td>\n",
       "      <td>c-438</td>\n",
       "      <td>tumor</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>F</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>breast</td>\n",
       "      <td>breast cancer</td>\n",
       "      <td>adenocarcinoma</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>adherent</td>\n",
       "      <td>MCF7_BREAST</td>\n",
       "      <td>IBMF-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>PC3</td>\n",
       "      <td>CVCL_0035</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72</td>\n",
       "      <td>F-12K ATCC catalog # 3-24</td>\n",
       "      <td>CRL-1435</td>\n",
       "      <td>c-214</td>\n",
       "      <td>tumor</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>M</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>prostate</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>adenocarcinoma</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>mix</td>\n",
       "      <td>PC3_PROSTATE</td>\n",
       "      <td>PC.3|PC-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>VCAP</td>\n",
       "      <td>CVCL_2235</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>220</td>\n",
       "      <td>DMEM ATCC catalog # 3-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tumor</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>M</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>prostate</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>adenocarcinoma</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>adherent</td>\n",
       "      <td>VCAP_PROSTATE</td>\n",
       "      <td>Vcap</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell_iname cellosaurus_id donor_age  donor_age_death  \\\n",
       "8      HEK293T            NaN       NaN              NaN   \n",
       "97        HA1E            NaN       NaN              NaN   \n",
       "113       A549      CVCL_0023        58              NaN   \n",
       "129       A375      CVCL_0132        54              NaN   \n",
       "138       HT29      CVCL_0320        44              NaN   \n",
       "151       MCF7      CVCL_0031        40              NaN   \n",
       "165        PC3      CVCL_0035        62              NaN   \n",
       "167       VCAP      CVCL_2235       NaN              NaN   \n",
       "\n",
       "     donor_disease_age_onset doubling_time  \\\n",
       "8                        NaN           NaN   \n",
       "97                       NaN            60   \n",
       "113                      NaN            48   \n",
       "129                      NaN            36   \n",
       "138                      NaN            36   \n",
       "151                      NaN            72   \n",
       "165                      NaN            72   \n",
       "167                      NaN           220   \n",
       "\n",
       "                                         growth_medium provider_catalog_id  \\\n",
       "8                                                  NaN                 NaN   \n",
       "97   MEM-ALPHA (Invitrogen A1049001) supplemented w...                 NaN   \n",
       "113                          F-12K ATCC catalog # 3-24             CCL-185   \n",
       "129                 DMEM Invitrogen catalog # 11995-65            CRL-1619   \n",
       "138             McCoy's 5A Invitrogen catalog # 166-82              HTB-38   \n",
       "151                           EMEM ATCC catalog # 3-23              HTB-22   \n",
       "165                          F-12K ATCC catalog # 3-24            CRL-1435   \n",
       "167                           DMEM ATCC catalog # 3-22                 NaN   \n",
       "\n",
       "    feature_id cell_type donor_ethnicity donor_sex donor_tumor_phase  \\\n",
       "8          NaN    normal         Unknown   Unknown           Unknown   \n",
       "97         NaN    normal         Unknown   Unknown           Unknown   \n",
       "113        c-4     tumor       Caucasian         M           Primary   \n",
       "129      c-127     tumor         Unknown         F        Metastatic   \n",
       "138      c-272     tumor       Caucasian         F           Primary   \n",
       "151      c-438     tumor       Caucasian         F        Metastatic   \n",
       "165      c-214     tumor       Caucasian         M        Metastatic   \n",
       "167        NaN     tumor       Caucasian         M        Metastatic   \n",
       "\n",
       "        cell_lineage       primary_disease                   subtype  \\\n",
       "8            unknown               unknown                   unknown   \n",
       "97            kidney  normal kidney sample      normal kidney sample   \n",
       "113             lung           lung cancer  non small cell carcinoma   \n",
       "129             skin           skin cancer                  melanoma   \n",
       "138  large_intestine          colon cancer            adenocarcinoma   \n",
       "151           breast         breast cancer            adenocarcinoma   \n",
       "165         prostate       prostate cancer            adenocarcinoma   \n",
       "167         prostate       prostate cancer            adenocarcinoma   \n",
       "\n",
       "    provider_name growth_pattern             ccle_name   cell_alias  \n",
       "8             NaN        unknown                   NaN         293T  \n",
       "97            NaN        unknown           HA1E_KIDNEY          NaN  \n",
       "113          ATCC       adherent             A549_LUNG        A 549  \n",
       "129          ATCC       adherent             A375_SKIN  A 375|A-375  \n",
       "138          ATCC       adherent  HT29_LARGE_INTESTINE        HT 29  \n",
       "151          ATCC       adherent           MCF7_BREAST       IBMF-7  \n",
       "165          ATCC            mix          PC3_PROSTATE    PC.3|PC-3  \n",
       "167          ATCC       adherent         VCAP_PROSTATE         Vcap  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# looking for if any healthy counterpart existed for the cell lines we selected for oe\n",
    "#  ['A549', 'PC3', 'A375', 'HEK293T', 'HA1E', 'MCF7', 'HT29', 'VCAP']\n",
    "cell_line_info = pd.read_csv('/home/b-evelyntong/hl/PDGrapher/data/raw/lincs/2022-02-LINCS_Level3/data/cellinfo_beta.txt', sep='\\t')\n",
    "cell_line_info[cell_line_info['cell_iname'].isin( ['A549', 'PC3', 'A375', 'HEK293T', 'HA1E', 'MCF7', 'HT29', 'VCAP'])]\n",
    "\n",
    "# here I realize that HEK293T and HA1E are normal tissues, and decide to keep them -- could be interesting to see\n",
    "\n",
    "# A375 and HT29 do not have healthy counterparts\n",
    "# A549 has known healthy counterpart NL20\n",
    "# PC3 and VCAP has known healthy RWPE1 ... even though it says it's also tumor but gonzalez used it so\n",
    "# MCF7 has known healthy MCF10A\n",
    "\n",
    "# so for the pt export loop, we have:\n",
    "# for cell_line, healthy in zip(['A549', 'MCF7', 'PC3', 'VCAP'], [('NL20', 'ctl_vehicle'), ('MCF10A', 'ctl_untrt'), ('RWPE1', 'ctl_vector'), ('RWPE1', 'ctl_vector')]):\n",
    "# for cell_line, healthy in zip(['A375', 'HT29', 'HEK293T', 'HA1E'], [None, None, None, None]):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of observational datapoints:\t4327\n",
      "\n",
      "Number of interventional datapoints:\t14047\n",
      "\n",
      "Number of interventional datapoints after keeping only those with perturbed genes in PPI:\t12642\n",
      "\n",
      "Assembling data...\n",
      "\n",
      "Samples forward:\t0\n",
      "\n",
      "Samples backward:\t12642\n",
      "\n",
      "Unique perturbagens:\t1390\n",
      "\n",
      "Saving data A549 ...\n",
      "\n",
      "\n",
      "\n",
      "Number of observational datapoints:\t4852\n",
      "\n",
      "Number of interventional datapoints:\t5464\n",
      "\n",
      "Number of interventional datapoints after keeping only those with perturbed genes in PPI:\t4495\n",
      "\n",
      "Assembling data...\n",
      "\n",
      "Samples forward:\t0\n",
      "\n",
      "Samples backward:\t4495\n",
      "\n",
      "Unique perturbagens:\t1473\n",
      "\n",
      "Saving data MCF7 ...\n",
      "\n",
      "\n",
      "\n",
      "Number of observational datapoints:\t6890\n",
      "\n",
      "Number of interventional datapoints:\t10192\n",
      "\n",
      "Number of interventional datapoints after keeping only those with perturbed genes in PPI:\t8751\n",
      "\n",
      "Assembling data...\n",
      "\n",
      "Samples forward:\t0\n",
      "\n",
      "Samples backward:\t8751\n",
      "\n",
      "Unique perturbagens:\t1692\n",
      "\n",
      "Saving data PC3 ...\n",
      "\n",
      "\n",
      "\n",
      "Number of observational datapoints:\t4184\n",
      "\n",
      "Number of interventional datapoints:\t4239\n",
      "\n",
      "Number of interventional datapoints after keeping only those with perturbed genes in PPI:\t3474\n",
      "\n",
      "Assembling data...\n",
      "\n",
      "Samples forward:\t0\n",
      "\n",
      "Samples backward:\t3474\n",
      "\n",
      "Unique perturbagens:\t1212\n",
      "\n",
      "Saving data VCAP ...\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 4/6\n",
    "# since we don't have cosmic data file yet, I'll treat these cell lines as having no healthy counterparts for now so I can continue to test\n",
    "\n",
    "data_root_dir = \"/home/b-evelyntong/hl/lincs_lvl3_oe/processed\"\n",
    "# for cell_line, healthy in zip(['A549', 'MCF7', 'PC3', 'VCAP'], [('NL20', 'ctl_vehicle'), ('MCF10A', 'ctl_untrt'), ('RWPE1', 'ctl_vector'), ('RWPE1', 'ctl_vector')]):\n",
    "for cell_line, healthy in zip(['A549', 'MCF7', 'PC3', 'VCAP'], [None, None, None, None]):\n",
    "# load_ppi\n",
    "\tppi = nx.read_edgelist(\"/home/b-evelyntong/hl/PDGrapher/data/processed/ppi/ppi_all_genes_edgelist.txt\")\n",
    "\n",
    "# load_gene_metadata (supposedly only gene_info is used)\n",
    "\tgene_info = pd.read_csv(\"/home/b-evelyntong/hl/PDGrapher/data/processed/lincs/gene_info.txt\")\n",
    "\tdict_entrez_symbol = dict(zip(gene_info['gene_id'], gene_info['gene_symbol']))\n",
    "\tdict_symbol_entrez = dict(zip(gene_info['gene_symbol'], gene_info['gene_id']))\n",
    "\n",
    "################################ uncomment when I get cosmic ######################################\n",
    "# # load_healthy_data\n",
    "# \thealthy_data_path = osp.join(data_root_dir, 'cell_line_{}_pert_{}.npz'.format(healthy[0], healthy[1]))\n",
    "# \thealthy_metadata_path = osp.join(data_root_dir, 'cell_line_{}_pert_{}_metadata.txt'.format(healthy[0], healthy[1]))\n",
    "# \t#Loads metadata\n",
    "# \thealthy_metadata = pd.read_csv(healthy_metadata_path)\n",
    "# \t#Loads data\n",
    "# \twith np.load(healthy_data_path, allow_pickle=True) as arr:\n",
    "# \t\thealthy_data =arr['data']\n",
    "# \t\tcol_ids = arr['col_ids']\n",
    "# \t\trow_ids = arr['row_ids']\n",
    "# \thealthy_data = pd.DataFrame(healthy_data, columns= col_ids, index=row_ids)\n",
    "# \tprint('Loading healthy cell line:\\t{} Number of samples:\\t{}\\n'.format(healthy[0], healthy_data.shape[1]))\n",
    "\n",
    "# # load_cosmic\n",
    "# \tdata = pd.read_csv('CosmicCLP_MutantExport_only_verified_and_curated')\n",
    "# \tprint('Loading COSMIC data. Number of cell lines:\\t{}\\n'.format(len(set(data['Sample name']))))\n",
    "# # map_cosmic_to_lincs\n",
    "# \tcosmic_data = cosmic_data[cosmic_data['Sample name']==cell_line]\n",
    "# \tprint('Mapping cosmic genes to lincs. Mapped: {}/{}\\n'.format(len(set(cosmic_data['Gene name']).intersection(gene_info['gene_symbol'])), len(set(cosmic_data['Gene name']))))\n",
    "# \t#Filter genes not mapped to LINCS\n",
    "# \tcosmic_data = cosmic_data[[gene_symbol in dict_symbol_entrez for gene_symbol in cosmic_data['Gene name']]]\n",
    "# \t#Save COSMIC mutations as entrez id (dataframe index)\n",
    "# \ttry:\n",
    "# \t\tcosmic_mutations = list(set([dict_symbol_entrez[symbol] for symbol in cosmic_data['Gene name'].tolist()]))\n",
    "# \texcept:\n",
    "# \t\timport pdb; pdb.set_trace()\n",
    "\thealthy_data = None\n",
    "\tcosmic_mutations = None\n",
    "#############################################################################################\n",
    "\t\n",
    "# load_data\n",
    "\t#Loads data matrix (observational)\n",
    "\tfile = osp.join(data_root_dir, 'cell_line_{}_pert_ctl_vector.npz'.format(cell_line))\n",
    "\tfile_metadata = osp.join(data_root_dir, 'cell_line_{}_pert_ctl_vector_metadata.txt'.format(cell_line))\n",
    "\tobs_metadata = pd.read_csv(file_metadata)\n",
    "\twith np.load(file, allow_pickle=True) as arr:\n",
    "\t\tobs_data =arr['data']\n",
    "\t\tcol_ids = arr['col_ids']\n",
    "\t\trow_ids = arr['row_ids']\n",
    "\tobs_data = pd.DataFrame(obs_data, columns= col_ids, index=row_ids)\n",
    "\tprint('Number of observational datapoints:\\t{}\\n'.format(len(obs_metadata)))\n",
    "\t#Loads data matrix (interventional)\n",
    "\tfile = osp.join(data_root_dir, 'cell_line_{}_pert_trt_oe.npz'.format(cell_line))\n",
    "\tfile_metadata = osp.join(data_root_dir, 'cell_line_{}_pert_trt_oe_metadata.txt'.format(cell_line))\n",
    "\tint_metadata = pd.read_csv(file_metadata)\n",
    "\twith np.load(file, allow_pickle=True) as arr:\n",
    "\t\tint_data =arr['data']\n",
    "\t\tcol_ids = arr['col_ids']\n",
    "\t\trow_ids = arr['row_ids']\n",
    "\tint_data = pd.DataFrame(int_data, columns= col_ids, index=row_ids)\n",
    "\tprint('Number of interventional datapoints:\\t{}\\n'.format(len(int_metadata)))\n",
    "# filter_data\n",
    "\tcosmic_mutations = None\n",
    "\t#1.Filter out obs and int data to keep only genes that are in the PPI\n",
    "\tgene_symbols_in_ppi = list(ppi.nodes())\n",
    "\t#####\n",
    "\t# this code was buggy for keying non-existent gene in ppi\n",
    "\t# BECAUSE I SHORT CUTTED THE LOOP and gene_info was modified in place LOL\n",
    "\tdict_symbol_id = dict(zip(gene_info['gene_symbol'], gene_info['gene_id']))\n",
    "\tgene_ids_in_ppi = [dict_symbol_id[i] for i in gene_symbols_in_ppi]\n",
    "\tgene_info.index = gene_info['gene_id']; gene_info = gene_info.loc[gene_ids_in_ppi].reset_index(inplace=False, drop=True)\n",
    "\tobs_data = obs_data.loc[gene_ids_in_ppi]\n",
    "\tint_data = int_data.loc[gene_ids_in_ppi]\n",
    "\t######\n",
    "\tif healthy_data is not None:\n",
    "\t\thealthy_data = healthy_data.loc[gene_ids_in_ppi]\n",
    "\tif cosmic_mutations is not None:\n",
    "\t\tcosmic_mutations = pd.DataFrame(cosmic_mutations)[[e in gene_ids_in_ppi for e in cosmic_mutations]][0].tolist()\n",
    "\t#2. Filter out samples whose interventions are not in the remaining genes (those in the PPI)\n",
    "\tkeep = []\n",
    "\tfor i, gene_symbol in enumerate(int_metadata['cmap_name']):\n",
    "\t\tif gene_symbol in gene_symbols_in_ppi:\n",
    "\t\t\tkeep.append(int_metadata.at[i, 'sample_id'])\n",
    "\tint_metadata.index = int_metadata['sample_id']; int_metadata = int_metadata.loc[keep].reset_index(inplace=False, drop=True)\n",
    "\tint_data = int_data[keep]\n",
    "\tprint('Number of interventional datapoints after keeping only those with perturbed genes in PPI:\\t{}\\n'.format(len(int_metadata)))\n",
    "\tprint('Assembling data...\\n')\n",
    "\n",
    "\t#First, we re-index genes in PPI and data\n",
    "\t\t#Gene symbol to index to ordered index\n",
    "\tgene_symbol_to_index = dict(zip(gene_info['gene_symbol'], gene_info['gene_id']))\n",
    "\tgene_index_to_ordered_index = dict(zip(gene_info['gene_id'], range(len(gene_info))))\n",
    "\tgene_info['ordered_index'] = [gene_index_to_ordered_index[i] for i in gene_info['gene_id']]\n",
    "\n",
    "\n",
    "\t\t#Reindex genes in PPI, obs_data, int_data, healty_data, and cosmic_mutations\n",
    "\tppi = nx.relabel_nodes(ppi, gene_symbol_to_index)\n",
    "\tppi = nx.relabel_nodes(ppi, gene_index_to_ordered_index)\n",
    "\tint_data.index = [gene_index_to_ordered_index[i] for i in int_data.index]\n",
    "\tint_data = int_data.sort_index(inplace=False)\n",
    "\tobs_data.index = [gene_index_to_ordered_index[i] for i in obs_data.index]\n",
    "\tobs_data = obs_data.sort_index(inplace=False)\n",
    "\tif healthy_data is not None:\n",
    "\t\thealthy_data.index = [gene_index_to_ordered_index[i] for i in healthy_data.index]\n",
    "\t\thealthy_data = healthy_data.sort_index(inplace=False)\n",
    "\tif cosmic_mutations is not None:\n",
    "\t\tcosmic_mutations = [gene_index_to_ordered_index[i] for i in cosmic_mutations]\n",
    "\t\tcosmic_vector = np.zeros(len(healthy_data))\n",
    "\t\tcosmic_vector[cosmic_mutations] = 1\n",
    "\n",
    "\t#Assembling samples\n",
    "\tedge_index = torch.LongTensor(np.array(ppi.edges()).transpose())\n",
    "\tedge_index = add_remaining_self_loops(edge_index)[0]\n",
    "\tedge_index = to_undirected(edge_index)\n",
    "\tnumber_of_nodes = ppi.number_of_nodes()\n",
    "\n",
    "\t#remove incoming edges to perturbed nodes (mutated nodes)\n",
    "\t# mask = [e not in cosmic_mutations for e in edge_index[1,:]]\n",
    "\t# edge_index_mutilated = edge_index[:, mask]\n",
    "\t# edge_index_mutilated = add_remaining_self_loops(edge_index_mutilated)[0]\n",
    "\n",
    "\tdict_forward_sample_and_mutations = dict() #saves the mutation vector used in forward\n",
    "\n",
    "\tforward_data_list = []\n",
    "\tif healthy_data is not None: #we only build forward data when we have healthy samples\n",
    "\t\t#FORWARD DATA - healthy_data, cosmic_vector, obs_data\n",
    "\t\t\t#each Data object will be a pairing of a random healthy_data column, the cosmic mutations, and a obs_data column\n",
    "\t\t\t#will have as many as obs_data columns\n",
    "\t\ti = 0\n",
    "\t\torder = np.array(range(healthy_data.shape[1]))\n",
    "\t\tnp.random.shuffle(order)\n",
    "\t\tfor sample_id in obs_data.columns:\n",
    "\t\t\t#sample a random healthy GE vector\n",
    "\t\t\ti = i % healthy_data.shape[1]\n",
    "\t\t\tsample_index = order[i]\n",
    "\t\t\thealthy_sample = healthy_data[healthy_data.columns[i]].values\n",
    "\t\t\thealthy = torch.Tensor(healthy_sample)\n",
    "\t\t\t#mutation\n",
    "\t\t\t#randomize mutations. First select the percentage of mutations to include, then select the mutations\n",
    "\t\t\tperc_to_include = np.random.choice([0.25, 0.50, 0.75, 1], 1).item()\n",
    "\t\t\tif int_metadata['cell_mfc_name'][0].split('.')[0] == 'PC3':\n",
    "\t\t\t\tperc_to_include = 1\n",
    "\t\t\tcosmic_mutations_i = np.random.choice(cosmic_mutations, int(len(cosmic_mutations)* perc_to_include))\n",
    "\t\t\tcosmic_vector = np.zeros(len(healthy_data))\n",
    "\t\t\tcosmic_vector[cosmic_mutations_i] = 1\n",
    "\t\t\tmutations = torch.Tensor(cosmic_vector)\n",
    "\t\t\t#diseased\n",
    "\t\t\tdiseased = torch.Tensor(obs_data[sample_id])\n",
    "\t\t\tdata = Data(healthy = healthy, mutations=mutations, diseased=diseased, gene_symbols = gene_info['gene_symbol'].tolist())\n",
    "\t\t\tdata.num_nodes = number_of_nodes\n",
    "\t\t\tforward_data_list.append(data)\n",
    "\t\t\t#Save \n",
    "\t\t\ti +=1\n",
    "\t\t\tdict_forward_sample_and_mutations[sample_id] = mutations\n",
    "\t\tprint('finished data forward')\n",
    "\n",
    "\n",
    "\t#BACKWARD DATA - obs_data, int_data\n",
    "\t#dict sample id: perturbed gene ordered index\n",
    "\tdict_sample_id_perturbed_gene_ordered_index = dict()\n",
    "\tfor sample_id, cmap_name in zip(int_metadata['sample_id'], int_metadata['cmap_name']):\n",
    "\t\tdict_sample_id_perturbed_gene_ordered_index[sample_id] = gene_index_to_ordered_index[gene_symbol_to_index[cmap_name]]\n",
    "\n",
    "\n",
    "\t#these are helpers to sample from obs_data\n",
    "\torder = np.array(range(obs_data.shape[1]))\n",
    "\tnp.random.shuffle(order)\n",
    "\ti = 0\n",
    "\t#shuffle obs data columns\n",
    "\tbackward_data_list = []\n",
    "\tunique_names_pert = set()\n",
    "\tfor sample_id in int_data.columns:\n",
    "\t\tbinary_indicator_perturbation = np.zeros(len(int_data))\n",
    "\t\tbinary_indicator_perturbation[dict_sample_id_perturbed_gene_ordered_index[sample_id]] = 1\n",
    "\t\t#Get a random pre-intervention sample\n",
    "\t\ti = i % obs_data.shape[1]\n",
    "\t\tsample_index = order[i]\n",
    "\t\tobs_sample_id = obs_data.columns[i]\n",
    "\t\tobs_sample = obs_data[obs_data.columns[i]].values\n",
    "\t\t#concat initial node features and perturbation indicator\n",
    "\t\tdiseased = torch.Tensor(obs_sample)\n",
    "\t\tintervention = torch.Tensor(binary_indicator_perturbation)\n",
    "\t\tif healthy_data is not None:\n",
    "\t\t\tmutations = dict_forward_sample_and_mutations[obs_sample_id]\n",
    "\t\telse:\n",
    "\t\t\tmutations = torch.Tensor(np.zeros(len(diseased)))\n",
    "\t\t# torch.Tensor(np.stack([obs_sample, binary_indicator_perturbation], 1))\n",
    "\t\t#post-intervention\n",
    "\t\ttreated = torch.Tensor(int_data[sample_id])\n",
    "\t\t#remove incoming edges to perturbed node\n",
    "\t\t# perturbed_node = dict_sample_id_perturbed_gene_ordered_index[sample_id]\n",
    "\t\t# edge_index_mutilated = edge_index[:, edge_index[1,:] != perturbed_node]\n",
    "\t\t\n",
    "\t\tgene_name = int_metadata[int_metadata['sample_id'] == sample_id]['cmap_name'].item()\n",
    "\t\tunique_names_pert.add(gene_name)\n",
    "\t\tdata = Data(perturbagen_name = gene_name, diseased = diseased, intervention=intervention, treated = treated, gene_symbols = gene_info['gene_symbol'].tolist(), mutations = mutations)\n",
    "\t\tdata.num_nodes = number_of_nodes\n",
    "\t\tbackward_data_list.append(data)\n",
    "\t\ti +=1\n",
    "\n",
    "\tprint('Samples forward:\\t{}\\n'.format(len(forward_data_list)))\n",
    "\tprint('Samples backward:\\t{}\\n'.format(len(backward_data_list)))\n",
    "\tprint('Unique perturbagens:\\t{}\\n'.format(len(unique_names_pert)))\n",
    "\n",
    "# save_data\n",
    "\tprint('Saving data {} ...\\n\\n\\n'.format(cell_line))\n",
    "\t\n",
    "\toutdir = '/home/b-evelyntong/hl/lincs_lvl3_oe/pt'\n",
    "\ttorch.save(forward_data_list, osp.join(outdir, 'data_forward_{}.pt'.format(cell_line)))\n",
    "\ttorch.save(backward_data_list, osp.join(outdir, 'data_backward_{}.pt'.format(cell_line)))\n",
    "\ttorch.save(edge_index, osp.join(outdir, 'edge_index_{}.pt'.format(cell_line)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_root_dir = \"/home/b-evelyntong/hl/lincs_lvl3_oe/processed\"\n",
    "for cell_line, healthy in zip(['A375', 'HT29', 'HEK293T', 'HA1E'], [None, None, None, None]):\n",
    "# load_ppi\n",
    "\tppi = nx.read_edgelist(\"/home/b-evelyntong/hl/PDGrapher/data/processed/ppi/ppi_all_genes_edgelist.txt\")\n",
    "\n",
    "# load_gene_metadata (supposedly only gene_info is used)\n",
    "\tgene_info = pd.read_csv(\"/home/b-evelyntong/hl/PDGrapher/data/processed/lincs/gene_info.txt\")\n",
    "\t\n",
    "# load_data\n",
    "\t#Loads data matrix (observational)\n",
    "\tfile = osp.join(data_root_dir, 'cell_line_{}_pert_ctl_vector.npz'.format(cell_line))\n",
    "\tfile_metadata = osp.join(data_root_dir, 'cell_line_{}_pert_ctl_vector_metadata.txt'.format(cell_line))\n",
    "\tobs_metadata = pd.read_csv(file_metadata)\n",
    "\twith np.load(file, allow_pickle=True) as arr:\n",
    "\t\tobs_data =arr['data']\n",
    "\t\tcol_ids = arr['col_ids']\n",
    "\t\trow_ids = arr['row_ids']\n",
    "\tobs_data = pd.DataFrame(obs_data, columns= col_ids, index=row_ids)\n",
    "\tprint('Number of observational datapoints:\\t{}\\n'.format(len(obs_metadata)))\n",
    "\t#Loads data matrix (interventional)\n",
    "\tfile = osp.join(data_root_dir, 'cell_line_{}_pert_trt_oe.npz'.format(cell_line))\n",
    "\tfile_metadata = osp.join(data_root_dir, 'cell_line_{}_pert_trt_oe_metadata.txt'.format(cell_line))\n",
    "\tint_metadata = pd.read_csv(file_metadata)\n",
    "\twith np.load(file, allow_pickle=True) as arr:\n",
    "\t\tint_data =arr['data']\n",
    "\t\tcol_ids = arr['col_ids']\n",
    "\t\trow_ids = arr['row_ids']\n",
    "\tint_data = pd.DataFrame(int_data, columns= col_ids, index=row_ids)\n",
    "\tprint('Number of interventional datapoints:\\t{}\\n'.format(len(int_metadata)))\n",
    "# filter_data\n",
    "\thealthy_data = None\n",
    "\thealthy_metadata = None\n",
    "\tcosmic_mutations = None\n",
    "\t#1.Filter out obs and int data to keep only genes that are in the PPI\n",
    "\tgene_symbols_in_ppi = list(ppi.nodes())\n",
    "\t#####\n",
    "\t# this code was buggy for keying non-existent gene in ppi\n",
    "\t# BECAUSE I SHORT CUTTED THE LOOP and gene_info was modified in place LOL\n",
    "\tdict_symbol_id = dict(zip(gene_info['gene_symbol'], gene_info['gene_id']))\n",
    "\tgene_ids_in_ppi = [dict_symbol_id[i] for i in gene_symbols_in_ppi]\n",
    "\tgene_info.index = gene_info['gene_id']; gene_info = gene_info.loc[gene_ids_in_ppi].reset_index(inplace=False, drop=True)\n",
    "\tobs_data = obs_data.loc[gene_ids_in_ppi]\n",
    "\tint_data = int_data.loc[gene_ids_in_ppi]\n",
    "\t######\n",
    "\tif healthy_data is not None:\n",
    "\t\thealthy_data = healthy_data.loc[gene_ids_in_ppi]\n",
    "\tif cosmic_mutations is not None:\n",
    "\t\tcosmic_mutations = pd.DataFrame(cosmic_mutations)[[e in gene_ids_in_ppi for e in cosmic_mutations]][0].tolist()\n",
    "\t#2. Filter out samples whose interventions are not in the remaining genes (those in the PPI)\n",
    "\tkeep = []\n",
    "\tfor i, gene_symbol in enumerate(int_metadata['cmap_name']):\n",
    "\t\tif gene_symbol in gene_symbols_in_ppi:\n",
    "\t\t\tkeep.append(int_metadata.at[i, 'sample_id'])\n",
    "\tint_metadata.index = int_metadata['sample_id']; int_metadata = int_metadata.loc[keep].reset_index(inplace=False, drop=True)\n",
    "\tint_data = int_data[keep]\n",
    "\tprint('Number of interventional datapoints after keeping only those with perturbed genes in PPI:\\t{}\\n'.format(len(int_metadata)))\n",
    "\tprint('Assembling data...\\n')\n",
    "\n",
    "\t#First, we re-index genes in PPI and data\n",
    "\t\t#Gene symbol to index to ordered index\n",
    "\tgene_symbol_to_index = dict(zip(gene_info['gene_symbol'], gene_info['gene_id']))\n",
    "\tgene_index_to_ordered_index = dict(zip(gene_info['gene_id'], range(len(gene_info))))\n",
    "\tgene_info['ordered_index'] = [gene_index_to_ordered_index[i] for i in gene_info['gene_id']]\n",
    "\n",
    "\n",
    "\t\t#Reindex genes in PPI, obs_data, int_data, healty_data, and cosmic_mutations\n",
    "\tppi = nx.relabel_nodes(ppi, gene_symbol_to_index)\n",
    "\tppi = nx.relabel_nodes(ppi, gene_index_to_ordered_index)\n",
    "\tint_data.index = [gene_index_to_ordered_index[i] for i in int_data.index]\n",
    "\tint_data = int_data.sort_index(inplace=False)\n",
    "\tobs_data.index = [gene_index_to_ordered_index[i] for i in obs_data.index]\n",
    "\tobs_data = obs_data.sort_index(inplace=False)\n",
    "\tif healthy_data is not None:\n",
    "\t\thealthy_data.index = [gene_index_to_ordered_index[i] for i in healthy_data.index]\n",
    "\t\thealthy_data = healthy_data.sort_index(inplace=False)\n",
    "\tif cosmic_mutations is not None:\n",
    "\t\tcosmic_mutations = [gene_index_to_ordered_index[i] for i in cosmic_mutations]\n",
    "\t\tcosmic_vector = np.zeros(len(healthy_data))\n",
    "\t\tcosmic_vector[cosmic_mutations] = 1\n",
    "\n",
    "\t#Assembling samples\n",
    "\tedge_index = torch.LongTensor(np.array(ppi.edges()).transpose())\n",
    "\tedge_index = add_remaining_self_loops(edge_index)[0]\n",
    "\tedge_index = to_undirected(edge_index)\n",
    "\tnumber_of_nodes = ppi.number_of_nodes()\n",
    "\n",
    "\t#remove incoming edges to perturbed nodes (mutated nodes)\n",
    "\t# mask = [e not in cosmic_mutations for e in edge_index[1,:]]\n",
    "\t# edge_index_mutilated = edge_index[:, mask]\n",
    "\t# edge_index_mutilated = add_remaining_self_loops(edge_index_mutilated)[0]\n",
    "\n",
    "\tdict_forward_sample_and_mutations = dict() #saves the mutation vector used in forward\n",
    "\n",
    "\tforward_data_list = []\n",
    "\tif healthy_data is not None: #we only build forward data when we have healthy samples\n",
    "\t\t#FORWARD DATA - healthy_data, cosmic_vector, obs_data\n",
    "\t\t\t#each Data object will be a pairing of a random healthy_data column, the cosmic mutations, and a obs_data column\n",
    "\t\t\t#will have as many as obs_data columns\n",
    "\t\ti = 0\n",
    "\t\torder = np.array(range(healthy_data.shape[1]))\n",
    "\t\tnp.random.shuffle(order)\n",
    "\t\tfor sample_id in obs_data.columns:\n",
    "\t\t\t#sample a random healthy GE vector\n",
    "\t\t\ti = i % healthy_data.shape[1]\n",
    "\t\t\tsample_index = order[i]\n",
    "\t\t\thealthy_sample = healthy_data[healthy_data.columns[i]].values\n",
    "\t\t\thealthy = torch.Tensor(healthy_sample)\n",
    "\t\t\t#mutation\n",
    "\t\t\t#randomize mutations. First select the percentage of mutations to include, then select the mutations\n",
    "\t\t\tperc_to_include = np.random.choice([0.25, 0.50, 0.75, 1], 1).item()\n",
    "\t\t\tif int_metadata['cell_mfc_name'][0].split('.')[0] == 'PC3':\n",
    "\t\t\t\tperc_to_include = 1\n",
    "\t\t\tcosmic_mutations_i = np.random.choice(cosmic_mutations, int(len(cosmic_mutations)* perc_to_include))\n",
    "\t\t\tcosmic_vector = np.zeros(len(healthy_data))\n",
    "\t\t\tcosmic_vector[cosmic_mutations_i] = 1\n",
    "\t\t\tmutations = torch.Tensor(cosmic_vector)\n",
    "\t\t\t#diseased\n",
    "\t\t\tdiseased = torch.Tensor(obs_data[sample_id])\n",
    "\t\t\tdata = Data(healthy = healthy, mutations=mutations, diseased=diseased, gene_symbols = gene_info['gene_symbol'].tolist())\n",
    "\t\t\tdata.num_nodes = number_of_nodes\n",
    "\t\t\tforward_data_list.append(data)\n",
    "\t\t\t#Save \n",
    "\t\t\ti +=1\n",
    "\t\t\tdict_forward_sample_and_mutations[sample_id] = mutations\n",
    "\t\tprint('finished data forward')\n",
    "\n",
    "\n",
    "\t#BACKWARD DATA - obs_data, int_data\n",
    "\t#dict sample id: perturbed gene ordered index\n",
    "\tdict_sample_id_perturbed_gene_ordered_index = dict()\n",
    "\tfor sample_id, cmap_name in zip(int_metadata['sample_id'], int_metadata['cmap_name']):\n",
    "\t\tdict_sample_id_perturbed_gene_ordered_index[sample_id] = gene_index_to_ordered_index[gene_symbol_to_index[cmap_name]]\n",
    "\n",
    "\n",
    "\t#these are helpers to sample from obs_data\n",
    "\torder = np.array(range(obs_data.shape[1]))\n",
    "\tnp.random.shuffle(order)\n",
    "\ti = 0\n",
    "\t#shuffle obs data columns\n",
    "\tbackward_data_list = []\n",
    "\tunique_names_pert = set()\n",
    "\tfor sample_id in int_data.columns:\n",
    "\t\tbinary_indicator_perturbation = np.zeros(len(int_data))\n",
    "\t\tbinary_indicator_perturbation[dict_sample_id_perturbed_gene_ordered_index[sample_id]] = 1\n",
    "\t\t#Get a random pre-intervention sample\n",
    "\t\ti = i % obs_data.shape[1]\n",
    "\t\tsample_index = order[i]\n",
    "\t\tobs_sample_id = obs_data.columns[i]\n",
    "\t\tobs_sample = obs_data[obs_data.columns[i]].values\n",
    "\t\t#concat initial node features and perturbation indicator\n",
    "\t\tdiseased = torch.Tensor(obs_sample)\n",
    "\t\tintervention = torch.Tensor(binary_indicator_perturbation)\n",
    "\t\tif healthy_data is not None:\n",
    "\t\t\tmutations = dict_forward_sample_and_mutations[obs_sample_id]\n",
    "\t\telse:\n",
    "\t\t\tmutations = torch.Tensor(np.zeros(len(diseased)))\n",
    "\t\t# torch.Tensor(np.stack([obs_sample, binary_indicator_perturbation], 1))\n",
    "\t\t#post-intervention\n",
    "\t\ttreated = torch.Tensor(int_data[sample_id])\n",
    "\t\t#remove incoming edges to perturbed node\n",
    "\t\t# perturbed_node = dict_sample_id_perturbed_gene_ordered_index[sample_id]\n",
    "\t\t# edge_index_mutilated = edge_index[:, edge_index[1,:] != perturbed_node]\n",
    "\t\t\n",
    "\t\tgene_name = int_metadata[int_metadata['sample_id'] == sample_id]['cmap_name'].item()\n",
    "\t\tunique_names_pert.add(gene_name)\n",
    "\t\tdata = Data(perturbagen_name = gene_name, diseased = diseased, intervention=intervention, treated = treated, gene_symbols = gene_info['gene_symbol'].tolist(), mutations = mutations)\n",
    "\t\tdata.num_nodes = number_of_nodes\n",
    "\t\tbackward_data_list.append(data)\n",
    "\t\ti +=1\n",
    "\n",
    "\tprint('Samples forward:\\t{}\\n'.format(len(forward_data_list)))\n",
    "\tprint('Samples backward:\\t{}\\n'.format(len(backward_data_list)))\n",
    "\tprint('Unique perturbagens:\\t{}\\n'.format(len(unique_names_pert)))\n",
    "\n",
    "# save_data\n",
    "\tprint('Saving data {} ...\\n\\n\\n'.format(cell_line))\n",
    "\t\n",
    "\toutdir = '/home/b-evelyntong/hl/lincs_lvl3_oe/pt'\n",
    "\ttorch.save(forward_data_list, osp.join(outdir, 'data_forward_{}.pt'.format(cell_line)))\n",
    "\ttorch.save(backward_data_list, osp.join(outdir, 'data_backward_{}.pt'.format(cell_line)))\n",
    "\ttorch.save(edge_index, osp.join(outdir, 'edge_index_{}.pt'.format(cell_line)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EOD 4/4\n",
    "- stuck at cosmos processing, can't find the tsv file for clp_mutation\n",
    "- need to actually train model, and generate my own script for the training"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4/5\n",
    "\n",
    "I have emailed Xiang Lin to see if he has any idea on the Cosmic data file. Meanwhile, I have to get a working model.\n",
    "\n",
    "Going through _models.py, train.py, _utils.py, found loss functions\n",
    "\n",
    "Next question:\n",
    "How is the perturbation information passed into the model?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4/6\n",
    "continue on understanding scripts\n",
    "actually no, let's train a model on xpr data and figure out how to evaluate on oe data \\\n",
    "started single_fold.py for A549 at 5:28, meanwhile I'll do some scripting \\\n",
    "finished at ~5:45, finished scriptig process_lincs_oe_npz.py at 6:58 \\\n",
    "started single_fold.py for MCF7 at 7:10, meanwhile I'll take a sht, finished at 7:45 \\\n",
    "I'm gonna train a model (1fold) on this oe data, let's do A549 first bc we have a model trained for the xpr version of this cell line already... need to first figure out how to generate splits on a single cell line \\\n",
    "\n",
    "starting 7:50 (meanwhile we will train a multifold)\n",
    "1. write script export_lincs_oe_for_torch_geometric.py, for now we assume all have no healthy counterparts DONE\n",
    "2. understand split generation DONE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this code just does not make sense to me, \n",
    "# 1) why is i used instead of sample_index? \n",
    "# 2) why stack by axis 1? are we making n by 2 matrix instead of a 2n by 1 vector?\n",
    "# 3) why is stacking commented out?\n",
    "\n",
    "# get a random pre-intervention sample\n",
    "i = i % obs_data.shape[1]\n",
    "sample_index = order[i]\n",
    "obs_sample_id = obs_data.columns[i]\n",
    "obs_sample = obs_data[obs_data.columns[i]].values  # an expression vector\n",
    "\n",
    "# concat initial node features and perturbation indicator\n",
    "diseased = torch.Tensor(obs_sample)\n",
    "intervention = torch.Tensor(binary_indicator_perturbation)\n",
    "mutations = torch.Tensor(np.zeros(len(diseased)))\n",
    "# torch.Tensor(np.stack([obs_sample, binary_indicator_perturbation], 1))#"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OMG the model is training on OE data!!!!!\n",
    "\n",
    "# 4/7\n",
    "Got cosmic processed file from Xiang, start torch export with forward at 3:26\n",
    "bugged out because I didn't process healthy metadata (there's a script for it)\n",
    "- process healthy metadata DONE\n",
    "- making models for each cell line from 1600 to 1900\n",
    "\n",
    "at 1800, /home/b-evelyntong/hl/lincs_lvl3_oe/torch_geometric_4_7 now has the complete, supposedly accurate data with both healthy and no health counterparts for the 8 selected oe cell lines\n",
    "\n",
    "at 2000, I now have 8 models each trained on one oe cell line... FK i forgot to turn forward data = true on the cell lines that now have forward data, so I'm gonna start this loop\n",
    "\n",
    "ok fk it, since I also need to create splits for the forward I'm just gonna re run the whole model loop\n",
    "\n",
    "- train_04071700: 8 cell lines, no forward data, 1 layer for each, 5 fold, 1 epoch\n",
    "- train_0407_small: 8 cell lines, with forward as apprpriate, 2 fold, 1 epoch, [\"n_layers_nn\": 1, \"n_layers_gnn\": 1]\n",
    "- train_0407_big: 8 cell lines, with forward as apprpriate, 5 fold, 5 epoch, [\"n_layers_nn\": 2, \"n_layers_gnn\": 2, \"positional_features_dim\": 64, \"embedding_layer_dim\": 8, \"dim_gnn\": 8]\n",
    "\n",
    "Goal for tonight: formally train big model for oe, get scripts automated for download data to train model for both xpr and oe (tmr morning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4/8\n",
    "0407_big got disconnected half way, made script to hopefully automate all downloading and processing"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pdgrapher",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
